File: frontend/convex/reagentCart.ts
import { query, mutation } from "./_generated/server";
import { v } from "convex/values";
import { requireAuth } from "./authHelper";

// ── List current user's cart items ────────────────────────────────────────────

export const list = query({
  args: {
    token: v.optional(v.string()),
    status: v.optional(v.string()),
  },
  handler: async (ctx, { token, status }) => {
    const userId = await requireAuth(ctx, token);

    let items = await ctx.db
      .query("reagent_cart")
      .withIndex("by_user", (q) => q.eq("user_id", userId))
      .collect();

    if (status) {
      items = items.filter((i) => i.status === status);
    }

    return items;
  },
});

// ── Create a cart item ────────────────────────────────────────────────────────

export const create = mutation({
  args: {
    token: v.optional(v.string()),
    name: v.string(),
    catalog_number: v.optional(v.string()),
    supplier: v.optional(v.string()),
    quantity: v.number(),
    unit: v.optional(v.string()),
    unit_price: v.optional(v.number()),
    url: v.optional(v.string()),
    notes: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const userId = await requireAuth(ctx, args.token);

    return await ctx.db.insert("reagent_cart", {
      user_id: userId,
      name: args.name,
      catalog_number: args.catalog_number,
      supplier: args.supplier,
      quantity: args.quantity,
      unit: args.unit,
      unit_price: args.unit_price,
      url: args.url,
      notes: args.notes,
      status: "pending",
      created_at: Date.now(),
    });
  },
});

// ── Update a cart item ────────────────────────────────────────────────────────

export const update = mutation({
  args: {
    token: v.optional(v.string()),
    id: v.id("reagent_cart"),
    name: v.optional(v.string()),
    catalog_number: v.optional(v.string()),
    supplier: v.optional(v.string()),
    quantity: v.optional(v.number()),
    unit: v.optional(v.string()),
    unit_price: v.optional(v.number()),
    url: v.optional(v.string()),
    notes: v.optional(v.string()),
    status: v.optional(v.string()),
  },
  handler: async (ctx, { token, id, ...fields }) => {
    const userId = await requireAuth(ctx, token);

    const item = await ctx.db.get(id);
    if (!item) throw new Error("Cart item not found");
    if (item.user_id !== userId) throw new Error("Not authorized to update this item");

    const patch: Record<string, unknown> = {};
    if (fields.name !== undefined) patch.name = fields.name;
    if (fields.catalog_number !== undefined) patch.catalog_number = fields.catalog_number;
    if (fields.supplier !== undefined) patch.supplier = fields.supplier;
    if (fields.quantity !== undefined) patch.quantity = fields.quantity;
    if (fields.unit !== undefined) patch.unit = fields.unit;
    if (fields.unit_price !== undefined) patch.unit_price = fields.unit_price;
    if (fields.url !== undefined) patch.url = fields.url;
    if (fields.notes !== undefined) patch.notes = fields.notes;
    if (fields.status !== undefined) patch.status = fields.status;

    await ctx.db.patch(id, patch);
    return id;
  },
});

// ── Remove a cart item ────────────────────────────────────────────────────────

export const remove = mutation({
  args: { token: v.optional(v.string()), id: v.id("reagent_cart") },
  handler: async (ctx, { token, id }) => {
    const userId = await requireAuth(ctx, token);

    const item = await ctx.db.get(id);
    if (!item) throw new Error("Cart item not found");
    if (item.user_id !== userId) throw new Error("Not authorized to remove this item");

    await ctx.db.delete(id);
    return id;
  },
});

// ── Checkout: mark selected items as ordered ──────────────────────────────────

export const checkout = mutation({
  args: {
    token: v.optional(v.string()),
    item_ids: v.array(v.id("reagent_cart")),
  },
  handler: async (ctx, { token, item_ids }) => {
    const userId = await requireAuth(ctx, token);

    const results: string[] = [];

    for (const id of item_ids) {
      const item = await ctx.db.get(id);
      if (!item) continue;
      if (item.user_id !== userId) continue; // Skip items not owned by the user

      await ctx.db.patch(id, { status: "ordered" });
      results.push(id);
    }

    return results;
  },
});

File: backend/app/api/reagent_cart.py
"""Reagent Cart API — receives captures from the browser extension and
serves them to the LabOS app. Designed to work whether or not Stripe
is configured."""
from datetime import datetime, timezone
from typing import Optional, List

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.models import ReagentCartItem, ReagentCartItemStatus, User
from app.services.auth import get_current_user

router = APIRouter(prefix="/reagent-cart", tags=["reagent-cart"])


class CartItemIn(BaseModel):
    vendor: str = ""
    name: str
    catalog: str = ""
    size: str = ""
    unit_price: Optional[float] = None
    quantity: int = 1
    currency: str = "USD"
    url: str = ""
    image_url: str = ""
    cas: str = ""
    purity: str = ""
    notes: str = ""
    captured_at: Optional[datetime] = None


class CartItemUpdate(BaseModel):
    quantity: Optional[int] = None
    notes: Optional[str] = None
    status: Optional[ReagentCartItemStatus] = None


class CheckoutIn(BaseModel):
    item_ids: List[int]
    payment_method_id: Optional[str] = None


def _to_out(item: ReagentCartItem) -> dict:
    return {
        "id": item.id,
        "vendor": item.vendor,
        "name": item.name,
        "catalog": item.catalog,
        "size": item.size,
        "unit_price": item.unit_price,
        "quantity": item.quantity,
        "currency": item.currency,
        "url": item.url,
        "image_url": item.image_url,
        "cas": item.cas,
        "purity": item.purity,
        "notes": item.notes,
        "status": item.status.value,
        "captured_at": item.captured_at.isoformat() if item.captured_at else None,
        "ordered_at": item.ordered_at.isoformat() if item.ordered_at else None,
    }


@router.get("/health")
def health():
    return {"ok": True}


@router.get("")
def list_items(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    q = db.query(ReagentCartItem).filter(
        (ReagentCartItem.user_id == user.id) | (ReagentCartItem.user_id.is_(None))
    ).order_by(ReagentCartItem.captured_at.desc()).limit(500)
    return [_to_out(i) for i in q.all()]


@router.post("", status_code=201)
def create_item(body: CartItemIn, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    item = ReagentCartItem(
        user_id=user.id,
        vendor=body.vendor,
        name=body.name,
        catalog=body.catalog,
        size=body.size,
        unit_price=body.unit_price,
        quantity=body.quantity,
        currency=body.currency,
        url=body.url,
        image_url=body.image_url,
        cas=body.cas,
        purity=body.purity,
        notes=body.notes,
        captured_at=body.captured_at or datetime.now(timezone.utc),
    )
    db.add(item)
    db.commit()
    db.refresh(item)
    return _to_out(item)


@router.patch("/{item_id}")
def update_item(item_id: int, body: CartItemUpdate, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    item = db.query(ReagentCartItem).filter(ReagentCartItem.id == item_id).first()
    if not item:
        raise HTTPException(404)
    if body.quantity is not None:
        item.quantity = body.quantity
    if body.notes is not None:
        item.notes = body.notes
    if body.status is not None:
        item.status = body.status
    db.commit()
    db.refresh(item)
    return _to_out(item)


@router.delete("/{item_id}", status_code=204)
def delete_item(item_id: int, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    item = db.query(ReagentCartItem).filter(ReagentCartItem.id == item_id).first()
    if not item:
        raise HTTPException(404)
    db.delete(item)
    db.commit()


@router.post("/checkout")
def checkout(body: CheckoutIn, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    """Mark items as ordered. If Stripe is configured, charge the saved
    payment method. Otherwise just mark the items as 'approved' and
    let the user process them externally."""
    items = db.query(ReagentCartItem).filter(ReagentCartItem.id.in_(body.item_ids)).all()
    if not items:
        raise HTTPException(404, "No items found for given IDs")

    total = sum((i.unit_price or 0) * (i.quantity or 1) for i in items)

    # Stripe path (graceful skip if not configured)
    import os
    stripe_key = os.environ.get("STRIPE_SECRET_KEY", "")
    charge_id = ""
    if stripe_key and body.payment_method_id:
        try:
            import stripe  # type: ignore
            stripe.api_key = stripe_key
            intent = stripe.PaymentIntent.create(
                amount=int(total * 100),
                currency="usd",
                payment_method=body.payment_method_id,
                confirm=True,
                off_session=True,
                description=f"LabOS reagent order: {len(items)} item(s)",
            )
            charge_id = intent.get("latest_charge") or intent.get("id", "")
        except Exception as e:
            raise HTTPException(402, f"Payment failed: {e}")

    now = datetime.now(timezone.utc)
    for it in items:
        it.status = ReagentCartItemStatus.ordered
        it.ordered_at = now
        if charge_id:
            it.stripe_charge_id = charge_id
    db.commit()

    return {
        "ok": True,
        "ordered_count": len(items),
        "total_amount": total,
        "stripe_charge_id": charge_id,
        "mode": "stripe" if charge_id else "manual",
    }
